File: NetFramework/XHtmlEditor/VB/RichTextEditor/Dialogs/SetLanguage.vb
Imports System.IO
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports System.Globalization
Imports C1.Win.C1SpellChecker

Namespace RichTextEditor
	Public Partial Class SetLanguage
		Inherits Form
		' ** fields
		Private _spellChecker As C1SpellChecker

		' ** ctors
		Public Sub New()
			InitializeComponent()
		End Sub
		Public Sub New(spellChecker As C1SpellChecker)
			InitializeComponent()

			' save reference to spell checker
			_spellChecker = spellChecker

			' populate list
			_list.Items.Add("Default (built-in English US)")
			For Each fn As String In Directory.GetFiles(Application.StartupPath, "c1spell_*.dct")
				Dim item As Dictionary = New Dictionary(fn)
				If item.IsValid Then
					_list.Items.Add(item)
				End If
			Next

			' select current
			_list.SelectedIndex = 0
			Dim current As String = _spellChecker.MainDictionary.FileName
			If Not String.IsNullOrEmpty(current) Then
				For Each item As Object In _list.Items
					Dim dct As Dictionary = TryCast(item, Dictionary)
					If dct IsNot Nothing AndAlso String.Equals(Path.GetFileName(dct.FileName), Path.GetFileName(current), StringComparison.OrdinalIgnoreCase) Then
						_list.SelectedItem = dct
						Exit For
					End If
				Next
			End If
		End Sub

		' ** event handlers
		Private Sub _btnOK_Click(sender As Object, e As EventArgs)
			' save main dictionary
			Dim dct As Dictionary = TryCast(_list.SelectedItem, Dictionary)
            _spellChecker.MainDictionary.FileName = Util.IIFS(dct IsNot Nothing, dct.FileName, Nothing)
		End Sub

		''' <summary>
		''' Helper class to keep track of dictionary names and languages.
		''' </summary>
		Private Class Dictionary
			Private _fileName As String
			Private _ci As CultureInfo

			Public Sub New(fileName As String)
				_fileName = fileName
				Dim name As String = Path.GetFileNameWithoutExtension(fileName).Substring(8)
				Try
					_ci = New CultureInfo(name)
				Catch
				End Try
            End Sub

			Public ReadOnly Property IsValid() As Boolean
				Get
					Return _ci IsNot Nothing
				End Get
            End Property

			Public ReadOnly Property FileName() As String
				Get
					Return _fileName
				End Get
            End Property

			Public Overloads Overrides Function ToString() As String
				Return _ci.EnglishName
			End Function
		End Class
	End Class
End Namespace

File: NetFramework/XHtmlEditor/CS/RichTextEditor/Dialogs/SetLanguage.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Globalization;
using C1.Win.C1SpellChecker;

namespace RichTextEditor
{
    public partial class SetLanguage : Form
    {
        // ** fields
        C1SpellChecker _spellChecker;

        // ** ctors
        public SetLanguage()
        {
            InitializeComponent();
        }
        public SetLanguage(C1SpellChecker spellChecker)
        {
            InitializeComponent();

            // save reference to spell checker
            _spellChecker = spellChecker;

            // populate list
            _list.Items.Add("Default (built-in English US)");
            foreach (string fn in Directory.GetFiles(Application.StartupPath, "c1spell_*.dct"))
            {
                Dictionary item = new Dictionary(fn);
                if (item.IsValid)
                {
                    _list.Items.Add(item);
                }
            }

            // select current
            _list.SelectedIndex = 0;
            string current = _spellChecker.MainDictionary.FileName;
            if (!string.IsNullOrEmpty(current))
            {
                foreach (Object item in _list.Items)
                {
                    Dictionary dct = item as Dictionary;
                    if (dct != null &&
                        string.Equals(Path.GetFileName(dct.FileName), Path.GetFileName(current), StringComparison.OrdinalIgnoreCase))
                    {
                        _list.SelectedItem = dct;
                        break;
                    }
                }
            }
        }

        // ** event handlers
        private void _btnOK_Click(object sender, EventArgs e)
        {
            // save main dictionary
            Dictionary dct = _list.SelectedItem as Dictionary;
            _spellChecker.MainDictionary.FileName = dct != null
                ? dct.FileName
                : null;
        }

        /// <summary>
        /// Helper class to keep track of dictionary names and languages.
        /// </summary>
        class Dictionary
        {
            string _fileName;
            CultureInfo _ci;

            public Dictionary(string fileName)
            {
                _fileName = fileName;
                string name = Path.GetFileNameWithoutExtension(fileName).Substring(8);
                try
                {
                    _ci = new CultureInfo(name);
                }
                catch { }
            }
            public bool IsValid
            {
                get { return _ci != null; }
            }
            public string FileName
            {
                get { return _fileName; }
            }
            public override string ToString()
            {
                return _ci.EnglishName;
            }
        }
    }
}
